### PR TITLE
[PLGN-194] Active Directory LDAP - Enable/Disable Users

### PR DESCRIPTION
**Pull request by ablakley-r7**
### Description

Describe the proposed changes:

  - Add two actions Enable Users and Disable Users
  - Accept a List input
  - Output successful operations and unsuccessful operations
  - [PLGN-194](https://issues.corp.rapid7.com/browse/PLGN-194)

## PR Requirements

Developers, verify you have completed the following items by checking them off:

#### Unit Tests

Review our documentation on [generating](https://docs.rapid7.com/insightconnect/unit-test-generation) and [writing](https://docs.rapid7.com/insightconnect/unit-test-primer) plugin unit tests

- [ ] Unit tests written for any new or updated code

#### In-Product Tests

If you are an InsightConnect customer or have access to an InsightConnect instance, the following in-product tests should be done:

- [ ] Screenshot of job output with the plugin changes
- [ ] Screenshot of the changed connection, actions, or triggers input within the InsightConnect workflow builder

### Style

Review the [style guide](https://docs.rapid7.com/insightconnect/style-guide/)

- [ ] For dependencies, pin [OS package](https://docs.rapid7.com/insightconnect/style-guide/#dockerfile) and [Python package](https://docs.rapid7.com/insightconnect/style-guide/#requirements.txt) versions
- [ ] For security, set least privileged account with ``USER nobody`` in the ``Dockerfile`` when possible
- [ ] For size, use the [slim SDK images](https://docs.rapid7.com/insightconnect/sdk-guide/#sdk-guide) when possible: ``komand/python-3-37-slim-plugin`` and ``komand/python-3-37-plugin``
- [ ] For error handling, use of [PluginException](https://docs.rapid7.com/insightconnect/error-handling-in-integrations/#plugin-exceptions) and [ConnectionTestException](https://docs.rapid7.com/insightconnect/error-handling-in-integrations#connection-exceptions)
- [ ] For logging, use [self.logger](https://docs.rapid7.com/insightconnect/sdk-guide/#logging)
- [ ] For docs, use [changelog style](https://docs.rapid7.com/insightconnect/style-guide/#changelog)
- [ ] For docs, validate markdown with ``make validate`` which calls ``mdl`` to lint ``help.md``

### Functional Checklist
- [ ] Work fully completed
- [ ] Functional
  - [ ] Any new actions/triggers include JSON [test files](https://docs.rapid7.com/insightconnect/style-guide/#tests) in the `tests/` directory created with `icon-plugin run -c sample $action > tests/$action.json`
  - [ ] Tests should all pass unless it's a negative test. Negative tests have a naming convention of `tests/$action_bad.json`
  - [ ] Unsuccessful tests should fail by raising an exception causing the plugin to die and an object should be returned on successful test
  - [ ] Add functioning test results to PR, sanitize any output if necessary
    * Single action/trigger `icon-plugin run -T tests/example.json --debug --jq`
    * All actions/triggers shortcut `icon-plugin run -T all --debug --jq` (use PR format at end)
  - [ ] Add functioning run results to PR, sanitize any output if necessary
    * Single action/trigger `icon-plugin run -R tests/example.json --debug --jq`
    * All actions/triggers shortcut `icon-plugin run -R all --debug --jq` (use PR format at end)

### Assessment

You must validate your work to reviewers:

1. Run `make validate` and make sure everything passes
2. Run the assessment tool: `icon-plugin run -A -R all -T all`. For single action validation: `icon-plugin run -A -R tests/my_action.json -T tests/my_action.json`
3. Copy (`icon-plugin ... | pbcopy`) and paste the output in **a new post** on this PR
4. Add required screenshots from the In-Product Tests section

### CODE REVIEW

File: plugins/active_directory_ldap/komand_active_directory_ldap/actions/disable_users/action.py [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+    def __init__(self):
+        super(self.__class__, self).__init__(
+            name="disable_users",
+            description=Component.DESCRIPTION,
+            input=DisableUsersInput(),
+            output=DisableUsersOutput(),
+        )
+
+    def run(self, params={}):
+        if len(params.get(Input.DISTINGUISHED_NAMES)) == 0:
```
Comment (cmcnally-r7) [inline plugins/active_directory_ldap/komand_active_directory_ldap/actions/disable_users/action.py:18]:
```suggestion
        if not len(params.get(Input.DISTINGUISHED_NAMES)):
```

Comment (cmcnally-r7) [reply plugins/active_directory_ldap/komand_active_directory_ldap/actions/disable_users/action.py:18]:
```
x = ""
len(x)
0
if not len(x):
    print(True)
    
True
if len(x):
    print(True)
```

File: plugins/active_directory_ldap/komand_active_directory_ldap/actions/disable_users/action.py [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+            raise PluginException(
+                cause="Distinguished Names must contain at least one entry",
+                assistance="Please enter one or more Distinguished Names",
+            )
+
+        success = True
+        disabled_users = self.connection.client.manage_users(params.get(Input.DISTINGUISHED_NAMES), False)
+        if len(disabled_users.get("successes")) == 0:
```
Comment (cmcnally-r7) [inline plugins/active_directory_ldap/komand_active_directory_ldap/actions/disable_users/action.py:26]: Same here

File: plugins/active_directory_ldap/komand_active_directory_ldap/actions/enable_users/action.py [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+                assistance="Please enter one or more Distinguished Names",
+            )
+
+        success = True
+        enabled_users = self.connection.client.manage_users(params.get(Input.DISTINGUISHED_NAMES), True)
+        if len(enabled_users.get("successes")) == 0:
+            success = False
+
+        return {
+            Output.SUCCESS: success,
```
Comment (igorski-r7) [inline plugins/active_directory_ldap/komand_active_directory_ldap/actions/enable_users/action.py:27]: Same here 😄 
```suggestion
        enabled_users = self.connection.client.manage_users(params.get(Input.DISTINGUISHED_NAMES), True)
        return {
            Output.SUCCESS: False if len(enabled_users.get("successes")) == 0 else True,
```

File: plugins/active_directory_ldap/komand_active_directory_ldap/actions/enable_users/action.py [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+    def run(self, params={}):
+        if len(params.get(Input.DISTINGUISHED_NAMES)) == 0:
+            raise PluginException(
+                cause="Distinguished Names must contain at least one entry",
+                assistance="Please enter one or more Distinguished Names",
+            )
+
+        success = True
+        enabled_users = self.connection.client.manage_users(params.get(Input.DISTINGUISHED_NAMES), True)
+        if len(enabled_users.get("successes")) == 0:
```
Comment (cmcnally-r7) [inline plugins/active_directory_ldap/komand_active_directory_ldap/actions/enable_users/action.py:23]: Same here

File: plugins/active_directory_ldap/komand_active_directory_ldap/actions/enable_users/action.py [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+
+
+class EnableUsers(insightconnect_plugin_runtime.Action):
+    def __init__(self):
+        super(self.__class__, self).__init__(
+            name="enable_users", description=Component.DESCRIPTION, input=EnableUsersInput(), output=EnableUsersOutput()
+        )
+
+    def run(self, params={}):
+        if len(params.get(Input.DISTINGUISHED_NAMES)) == 0:
```
Comment (cmcnally-r7) [inline plugins/active_directory_ldap/komand_active_directory_ldap/actions/enable_users/action.py:15]: Same here

File: plugins/active_directory_ldap/komand_active_directory_ldap/util/api.py [hunk 6]
```diff
@@ -165,6 +165,28 @@ def delete(self, conn, dn: str):
     def manage_user(self, conn, dn: str, status: bool):
         return ADUtils.change_account_status(conn, dn, status, self.logger)
 
+    @with_connection
+    def manage_users(self, conn, dns: List[str], status: bool) -> dict:
```
Comment (cmcnally-r7) [inline plugins/active_directory_ldap/komand_active_directory_ldap/util/api.py:169]: Can we type hint for the `conn` variable here?

File: plugins/active_directory_ldap/unit_test/test_action_disable_users.py
Comment (cmcnally-r7) [inline plugins/active_directory_ldap/unit_test/test_action_disable_users.py:1]: Is it possible to parameterise these tests

File: plugins/active_directory_ldap/unit_test/test_action_disable_users.py [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+            Output.SUCCESS: True,
+        }
+        self.assertEqual(expected, actual)
+
+    @default_connector(action=DisableUsers())
+    def test_empty_input(self, action):
+        with self.assertRaises(PluginException) as context:
+            action.run({Input.DISTINGUISHED_NAMES: []})
+        self.assertEqual("Distinguished Names must contain at least one entry", context.exception.cause)
+        self.assertEqual("Please enter one or more Distinguished Names", context.exception.assistance)
```
Comment (igorski-r7) [inline plugins/active_directory_ldap/unit_test/test_action_disable_users.py:64]: Could you please `parameterize` them? 😄

**Review by igorski-r7 — CHANGES_REQUESTED** [review]

File: plugins/active_directory_ldap/komand_active_directory_ldap/util/api.py (hunk 6 above)
Comment (ablakley-r7) [reply plugins/active_directory_ldap/komand_active_directory_ldap/util/api.py:169]: It's not explicitly used

File: plugins/active_directory_ldap/komand_active_directory_ldap/actions/disable_users/action.py [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+            success = False
+
+        return {
+            Output.SUCCESS: success,
```
[the remark this answers is not in the document]
Comment (ablakley-r7) [reply plugins/active_directory_ldap/komand_active_directory_ldap/actions/disable_users/action.py]: Validator hates this but I've updated based on your suggestion to make it shorter.

**Review by cmcnally-r7 — APPROVED** [review]

**Review by igorski-r7 — APPROVED** [review]

**Pull request merged**